pkg/controllers/pod/pod_controller.go

```diff
@@ -0,0 +1,310 @@
+// Responsible for monitoring kubernetes pods and
+// updating pod labels to calico ETCD if changed.
+type PodController struct {
+	indexer       cache.Indexer
+	informer      cache.Controller
+	labelCache    calicocache.ResourceCache
+	endpointCache map[string]api.WorkloadEndpoint
+	calicoClient  *client.Client
+	k8sClientset  *kubernetes.Clientset
+}
+
+// NewPodController Constructor for PodController
+func NewPodController(k8sClientset *kubernetes.Clientset, calicoClient *client.Client) controller.Controller {
+
+	podConverter := converter.NewPodConverter()
+
+	// Function returns map of podName:podObject stored by policy controller
+	// on ETCD datastore. Indentifies controller writen objects by
+	// their naming convention.
+	listFunc := func() (map[string]interface{}, error) {
+
+		filteredPods := make(map[string]interface{})
+
+		// Get all workloadEndpoints for kubernetes orchestrator from ETCD datastore
+
+		workloadEndpoints, err := calicoClient.WorkloadEndpoints().List(api.WorkloadEndpointMetadata{
+			Orchestrator: "k8s",
+		})
+		if err != nil {
+			return filteredPods, err
+		}
+
+		// Filter out only objects that are written by policy controller
+		for _, endpoint := range workloadEndpoints.Items {
+
+			key := podConverter.GetKey(endpoint)
+			filteredPods[key] = endpoint.Metadata.Labels
+		}
+		log.Debugf("Found %d pods in calico ETCD:", len(filteredPods))
+		return filteredPods, nil
+	}
+
+
+	cacheArgs := calicocache.ResourceCacheArgs{
+		ListFunc:   listFunc,
+		Client:     calicoClient,
+		ObjectType: reflect.TypeOf(map[string]string{}), // Restrict cache to store pod labels only.
+	}
+
+	labelCache := calicocache.NewResourceCache(cacheArgs)
+	endpointCache := make(map[string]api.WorkloadEndpoint)
+
+	
+
+	// create the watcher
+	listWatcher := cache.NewListWatchFromClient(k8sClientset.Core().RESTClient(), "pods", "", fields.Everything())
+
+	// Bind the calico cache to kubernetes cache with the help of an informer. This way we make sure that
+	// whenever the kubernetes cache is updated, changes get reflected in calico cache as well.
+	indexer, informer := cache.NewIndexerInformer(listWatcher, &v1.Pod{}, 0, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			log.Infof("Got ADD event for pod: %s\n", key)
+
+			endpoint, conversionErr := podConverter.Convert(obj)
+			if conversionErr != nil {
+				log.Errorf("Error while converting %#v to endpoint.", obj)
+				return
+			}
+
+			workloadEndpoint := endpoint.(api.WorkloadEndpoint)
+			calicoKey := podConverter.GetKey(workloadEndpoint)
+			// Add workloadID:Labels in labelCache
+			labelCache.Prime(calicoKey, workloadEndpoint.Metadata.Labels)
+		},
+		UpdateFunc: func(oldObj interface{}, newObj interface{}) {
+
+			key, err := cache.MetaNamespaceKeyFunc(newObj)
+
+			if err != nil {
+				log.Error(err)
```

Comment:
When logging errors, the idiomatic way is to use `WithError`.

So in this case:

```
log.WithError(err).Error("Failed to generate key")
```

Please make sure error logs are done this way throughout the codebase.